doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,336 @@
+
+- Per-Port link status changes  
+In this case, only the **link_training_status** will be updated.
+
+- Explicitly link training status refresh requests via the notification model.  
+In this case, the **link_training_failure** and **link_training_rxstatus** of the selected port will be refreshed. See detail description in section [Link Training Status Refresh](#link-training-status-refresh)
+
+Here is the table to map the fields and SAI attributes:  
+
+| **Parameter**          | **sai_port_attr_t**
+|:-----------------------|:-------------------------------------------|
+| link_training          | SAI_PORT_ATTR_LINK_TRAINING_ENABLE         |
+| link_training_failure  | SAI_PORT_ATTR_LINK_TRAINING_FAILURE_STATUS |
+| link_training_rxstatus | SAI_PORT_ATTR_LINK_TRAINING_RX_STATUS      |
+
+Please note the operational port link training status in the APPL_DB could be different from the administratively configured link training configuration in the CONFIG_DB due to the hardware limitations or vendor-specific SAI software implementations. For example, port link training is only applicable to the ports with speed >= 10G, and it's not applicable for chip-to-module transceivers.
+
+#### YANG Model Enhancements
+
+The port yang model needs to update according to DB schema change. The yang model changes of new fields are described below:
+
+```
+leaf link_training {
+  type string {
+    pattern "auto|on|off";
+  }
+}
+```
+
+These new yang leaf will be added to sonic-port.yang.
+
+#### DB Migration Considerations
+
+By having port link training disabled if "link_training" is not specified in the CONFIG_DB, this feature is fully backward compatible.
+
+#### SWSS Enhancements
+
+##### Port Configuration Flow
+
+The current SONiC port configuration flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+```
+
+The new SONiC port configuration flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if link_training changed and xcvr_capabilities has LT:
+    setPortLinkTraining(port, link_training)
+```
+
+##### Link Training Status Refresh
+
+To refresh the operational link training status in APPL_DB, a new notification handler will be introduced to PortsOrch, the pseudo code is as follow:
+
+```
+portCtrlConsumer = NotificationConsumer("PORT_CTRL")
+if portCtrlConsumer received 'op=port_status_refresh', 'data=EthernetXY' and 'type=link_training':
+    refreshPortLinkTrainingStatus(EthernetXY)
+```
+
+An example code snippet of initiating status refresh is as follow:
+
+```
+appl_db = daemon_base.db_connect("APPL_DB")
+port_np = swsscommon.NotificationProducer(appl_db, "PORT_CTRL")
+
+fvp = swsscommon.FieldValuePairs([
+    ('type', 'link_training')
+])
+port_np.send('port_state_refresh', port, fvp)
```

Comment:
Please add the sequence diagram. While portsorch is the consumer of this notification who is going to be the producer for this PORT_CTRL?

Reply:
Considering the further integration with CLIs (REST, KLISH...etc.) and avoid possible confusions, it's better to periodically poll the status when LT is enabled, hence this will be updated to a periodical polling